src/rebuild/initiator.c

```diff
@@ -870,17 +888,72 @@ rebuild_obj_ult(void *data)
 			break;
 	}
 
-	ds_obj_close(oh);
-free:
 	if (buf != NULL && buf != stack_buf)
 		D_FREE(buf);
+
+	D_DEBUG(DB_REBUILD, "obj "DF_UOID" for shard %u eph "
+		DF_U64": rc %d\n", DP_UOID(arg->oid), arg->shard,
+		epoch, rc);
+
+	return rc;
+}
+
+static int
+rebuild_obj_punch(struct rebuild_iter_obj_arg *arg)
+{
+	return dss_task_collective(rebuild_obj_punch_one, arg, 0);
+}
+
+/**
+ * Iterate akeys/dkeys of the object
+ */
+static void
+rebuild_obj_ult(void *data)
+{
+	struct rebuild_iter_obj_arg	*arg = data;
+	struct rebuild_pool_tls		*tls;
+	daos_handle_t			oh;
+	int				rc;
+
+	tls = rebuild_pool_tls_lookup(arg->rpt->rt_pool_uuid,
+				      arg->rpt->rt_rebuild_ver);
+	D_ASSERT(tls != NULL);
+
+	if (arg->epoch != DAOS_EPOCH_MAX) {
+		rc = rebuild_obj_punch(arg);
+		if (rc)
+			D_GOTO(free, rc);
+	}
+
+	rc = ds_obj_open(arg->cont_hdl, arg->oid.id_pub, DAOS_OO_RW, &oh);
+	if (rc)
+		D_GOTO(free, rc);
+
+	if (arg->snaps) {
+		int i;
+
+		for (i = 0; i < arg->snaps->snap_cnt; i++) {
+			rc = rebuild_one_epoch_object(oh, arg->snaps->snaps[i],
```

Comment:
for re-integration in future, is there a way to avoid rebuild already existed snapshot?

Reply:
Hmm, we need rebuild the object for all snapshots anyway. Actually not sure what you mean here?